File: Cpp/0303_range_sum_query_immutable.cpp
// https://leetcode.cn/problems/range-sum-query-immutable/

class NumArray {
public:
    NumArray(vector<int>& nums) {
        sum_to_now.resize(nums.size());
        int tmp = 0;
        for (int i = 0; i < nums.size(); ++i) {
            tmp = tmp + nums[i];
            sum_to_now[i] = tmp;
        }
    }
    
    int sumRange(int left, int right) {
        if (left == 0) {
            return sum_to_now[right];
        }
        else {
            return sum_to_now[right] - sum_to_now[left-1];
        }
    }

private:
    vector<int> sum_to_now;
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */

File: Python3/0303_range_sum_query_immutable.py
# https://leetcode.cn/problems/range-sum-query-immutable/

class NumArray:
    def __init__(self, nums):
        """
        Inputs:
            nums: list[int]
        """

        n = len(nums)
        self.nsum = [0 for i in range(n+1)]
        for i in range(n):
            self.nsum[i] = self.nsum[i-1] + nums[i]

    def sumRange(self, i, j):
        """
        Inputs:
            i: int
            j: int
        Outputs:
            res: int
        """

        res = self.nsum[j] - self.nsum[i-1]
        return res

# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(i,j)
